### PR TITLE
Move gpfdist to gpdb core.

### PR DESCRIPTION
**Pull request by Chibin**
As gpfdist will now be part of the gpdb build by default, we may just
treat it as part of gpdb core.
* Integrate gpfdist configure file to the top level configure and
changed necessary files to accomadate the changes.

### CODE REVIEW

**Comment by hlinnaka** [comment]
src/bin/gpfdist/configure should be removed

**Comment by hlinnaka** [comment]
I don't think this GP_VERSION ifdef is correct in gpfdist.c:

```
#ifndef GP_VERSION
#define GP_VERSION unknown
#endif
```

But in pg_config, we define GP_VERSION as:

```
#define GP_VERSION "4.3.99.00 build dev"
```

Which is, it should it include the quotes or not? (probably it should, and we should just remove the #ifdef, and rely on the #define in pg_config.h)

File: src/bin/gpfdist/Makefile [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+all: gpfdist$(EXE_EXT)
+
+fstream.c: $(top_builddir)/src/backend/utils/misc/fstream/fstream.c
+	ln -s $< $@
+
+gfile.c: $(top_builddir)/src/backend/utils/misc/fstream/gfile.c
+	ln -s $< $@
+
+gpfdist$(EXE_EXT): $(OBJS)
+	@if test "$(enable_gpfdist)" = yes; then \
```
Comment (hlinnaka) [inline src/bin/gpfdist/Makefile]: This enable_gpfdist test should be done in src/bin/Makefile instead, to avoid recursing into src/bin/gpfdist at all. At the moment, if you compile with --disable-gpfdist, we still compile all the .c files, and the only thing that's not performed is linking them into the gpfdist binary.

**Comment by hlinnaka** [comment]
I'm getting a ton of compiler warnings like this from transform.c:

```
transform.c:912:1: warning: no previous prototype for ‘transform_stderr_server’ [-Wmissing-prototypes]
 transform_stderr_server(struct transform* tr)
 ^
transform.c:918:1: warning: no previous prototype for ‘transform_content_paths’ [-Wmissing-prototypes]
 transform_content_paths(struct transform* tr)
 ^
transform.c:924:1: warning: no previous prototype for ‘transform_safe’ [-Wmissing-prototypes]
 transform_safe(struct transform* tr)
 ^
transform.c:930:1: warning: no previous prototype for ‘transform_saferegex’ [-Wmissing-prototypes]
 transform_saferegex(struct transform* tr)
```

**Comment by davecramer** [comment]
@xinzweb mentioned that gpfdist doesn't work on OSX, will this break OSX builds ?

**Comment by Chibin** [comment]
@davecramer No, this PR should "fix" the compiler problems that we previously had for OSX. 
I didn't test OSX build with clang and missed the -Werror flag on the configure for OSX (took it out), which is the only reason why it is failing - there will be tons of warning due to deprecated functions for OSX 10.7 if you use the default openssl on the OSX.

File: src/bin/gpfdist/Makefile (hunk 1 above)
Comment (Chibin) [inline src/bin/gpfdist/Makefile]: Ah, that's right... I'll fix it up.

**Comment by xinzweb** [comment]
@Chibin yeah, what's the current flag for other os? If they don't have `-Werror`, then we should NOT have it on OSX either.

**Comment by Chibin** [comment]
The current flag on master for all os is `-Werror`. It's just that OSX has deprecated functions and will complain about it.

**Comment by Chibin** [comment]
@hlinnaka  I'm trying to remove the have_glob check and use src/port/glob.c by default for all platform and I'm getting this error at the moment:
```
gcc -m64 -m64 -O3 -funroll-loops -fargument-noalias-global -fno-omit-frame-pointer -g -finline-limit=1800 -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-but-set-variable -Wno-address   -pthread -I/data/royc/tools/curl/7.44.0/dist/rhel5_x86_64/include -I../../../src/port -I. -I../../../src/include -D_GNU_SOURCE -I/usr/include/libxml2  -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/include -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/include/libxml2 -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/libgpopt/include -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/libgpos/include -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/addon/src/include  -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/include/apr-1   -DLINUX=2 -D_REENTRANT -D_GNU_SOURCE  -c -o ../../../src/port/glob.o ../../../src/port/glob.c
../../../src/port/glob.c: In function 'glob':
../../../src/port/glob.c:209: error: 'GLOB_MAGCHAR' undeclared (first use in this function)
../../../src/port/glob.c:209: error: (Each undeclared identifier is reported only once
../../../src/port/glob.c:209: error: for each function it appears in.)
../../../src/port/glob.c:215: error: 'GLOB_NOESCAPE' undeclared (first use in this function)
../../../src/port/glob.c:232: error: 'GLOB_BRACE' undeclared (first use in this function)
../../../src/port/glob.c: In function 'globtilde':
../../../src/port/glob.c:390: error: 'GLOB_TILDE' undeclared (first use in this function)
../../../src/port/glob.c: In function 'glob0':
../../../src/port/glob.c:496: error: 'GLOB_MAGCHAR' undeclared (first use in this function)
../../../src/port/glob.c:532: error: 'GLOB_NOMAGIC' undeclared (first use in this function)
../../../src/port/glob.c: In function 'glob3':
../../../src/port/glob.c:677: error: 'GLOB_ALTDIRFUNC' undeclared (first use in this function)
../../../src/port/glob.c: In function 'globextend':
../../../src/port/glob.c:774: error: 'GLOB_LIMIT' undeclared (first use in this function)
../../../src/port/glob.c: In function 'g_opendir':
../../../src/port/glob.c:867: error: 'GLOB_ALTDIRFUNC' undeclared (first use in this function)
../../../src/port/glob.c: In function 'g_lstat':
../../../src/port/glob.c:883: error: 'GLOB_ALTDIRFUNC' undeclared (first use in this function)
../../../src/port/glob.c: In function 'g_stat':
../../../src/port/glob.c:898: error: 'GLOB_ALTDIRFUNC' undeclared (first use in this function)
```

I took a look into the glob.h and found the following:
```
#ifndef _POSIX_SOURCE
#define GLOB_ALTDIRFUNC 0x0040  /* Use alternately specified directory funcs. */
#define GLOB_BRACE  0x0080  /* Expand braces ala csh. */
#define GLOB_MAGCHAR    0x0100  /* Pattern had globbing characters. */
#define GLOB_NOMAGIC    0x0200  /* GLOB_NOCHECK without magic chars (csh). */
#define GLOB_QUOTE  0x0400  /* Quote special chars with \. */
#define GLOB_TILDE  0x0800  /* Expand tilde names from the passwd file. */
#define GLOB_NOESCAPE   0x1000  /* Disable backslash escaping. */
#define GLOB_LIMIT  0x2000  /* Limit pattern match output to ARG_MAX */
#endif
```
Not entirely sure why it's this kind of check.
I would like to change it to the this ```#if !defined __USE_POSIX2 || defined __USE_BSD || defined __USE_GNU``` or just remove it entirely.

**Comment by yydzero** [comment]
If move to src/bin/gpfdist, do we still need --disable-gpfdist option?

On Wed, Feb 17, 2016 at 8:12 AM, Marbin Tan <notifications@github.com>
wrote:

> @hlinnaka <https://github.com/hlinnaka> I'm trying to remove the
> have_glob check and use src/port/glob.c by default for all platform and I'm
> getting this error at the moment:
>
> gcc -m64 -m64 -O3 -funroll-loops -fargument-noalias-global -fno-omit-frame-pointer -g -finline-limit=1800 -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-but-set-variable -Wno-address   -pthread -I/data/royc/tools/curl/7.44.0/dist/rhel5_x86_64/include -I../../../src/port -I. -I../../../src/include -D_GNU_SOURCE -I/usr/include/libxml2  -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/include -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/include/libxml2 -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/libgpopt/include -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/libgpos/include -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/s
>  rc/gpAux
>  /addon/src/include  -I/data/pulse2-agent/agents/agent1/work/GPDB-master-opt-l1-tanm82/rhel5_x86_64/src/gpAux/ext/rhel5_x86_64/include/apr-1   -DLINUX=2 -D_REENTRANT -D_GNU_SOURCE  -c -o ../../../src/port/glob.o ../../../src/port/glob.c
> ../../../src/port/glob.c: In function 'glob':
> ../../../src/port/glob.c:209: error: 'GLOB_MAGCHAR' undeclared (first use in this function)
> ../../../src/port/glob.c:209: error: (Each undeclared identifier is reported only once
> ../../../src/port/glob.c:209: error: for each function it appears in.)
> ../../../src/port/glob.c:215: error: 'GLOB_NOESCAPE' undeclared (first use in this function)
> ../../../src/port/glob.c:232: error: 'GLOB_BRACE' undeclared (first use in this function)
> ../../../src/port/glob.c: In function 'globtilde':
> ../../../src/port/glob.c:390: error: 'GLOB_TILDE' undeclared (first use in this function)
> ../../../src/port/glob.c: In function 'glob0':
> ../../../src/port/glob.c:496: error: 'GLOB_MAGCHAR' undeclared (first use in this function)
> ../../../src/port/glob.c:532: error: 'GLOB_NOMAGIC' undeclared (first use in this function)
> ../../../src/port/glob.c: In function 'glob3':
> ../../../src/port/glob.c:677: error: 'GLOB_ALTDIRFUNC' undeclared (first use in this function)
> ../../../src/port/glob.c: In function 'globextend':
> ../../../src/port/glob.c:774: error: 'GLOB_LIMIT' undeclared (first use in this function)
> ../../../src/port/glob.c: In function 'g_opendir':
> ../../../src/port/glob.c:867: error: 'GLOB_ALTDIRFUNC' undeclared (first use in this function)
> ../../../src/port/glob.c: In function 'g_lstat':
> ../../../src/port/glob.c:883: error: 'GLOB_ALTDIRFUNC' undeclared (first use in this function)
> ../../../src/port/glob.c: In function 'g_stat':
> ../../../src/port/glob.c:898: error: 'GLOB_ALTDIRFUNC' undeclared (first use in this function)
>
> I took a look into the glob.h and found the following:
>
> #ifndef _POSIX_SOURCE
> #define GLOB_ALTDIRFUNC 0x0040  /* Use alternately specified directory funcs. */
> #define GLOB_BRACE  0x0080  /* Expand braces ala csh. */
> #define GLOB_MAGCHAR    0x0100  /* Pattern had globbing characters. */
> #define GLOB_NOMAGIC    0x0200  /* GLOB_NOCHECK without magic chars (csh). */
> #define GLOB_QUOTE  0x0400  /* Quote special chars with \. */
> #define GLOB_TILDE  0x0800  /* Expand tilde names from the passwd file. */
> #define GLOB_NOESCAPE   0x1000  /* Disable backslash escaping. */
> #define GLOB_LIMIT  0x2000  /* Limit pattern match output to ARG_MAX */
> #endif
>
> Not entirely sure why it's this kind of check.
> I would like to change it to the this #if !defined __USE_POSIX2 ||
> defined __USE_BSD || defined __USE_GNU or just remove it entirely.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/351#issuecomment-184935371>.
>



-- 
Best Regards,
Yandong

**Comment by danielgustafsson** [comment]
I would like to keep the option, if only to be able to build faster during testing by being selective on what to include.

**Comment by Chibin** [comment]
> If move to src/bin/gpfdist, do we still need --disable-gpfdist option?

Yes, the disable option will still be available. I think it'd be useful for people who do not want it.

**Comment by Chibin** [comment]
@hlinnaka Please let me know if there's anything else that seems weird on this PR. Otherwise, I will be pushing this in soon. Thank you.

**Comment by hlinnaka** [comment]
Can you fix all the issues mentioned so far and update the PR, so that we can see what you're planning to push?

File: configure.in [hunk 2]
```diff
@@ -541,6 +548,7 @@ GP_VERSION_HEADER="src/include/catalog/gp_version.h"
 if grep '\$\$' $GP_VERSION_IN > /dev/null 2>&1 ; then
     sed "s,\\$.*\\$\\$,$GP_VERSION," $GP_VERSION_IN > $GP_VERSION_HEADER
 fi
+AC_SUBST(GP_VERSION)
```
Comment (hlinnaka) [inline configure.in]: I think it'd be cleaner to do #include "catalog/gp_version.h" in gpfdist and get this constant from there.

File: configure.in [hunk 3]
```diff
@@ -1147,6 +1191,7 @@ if test "$enable_snmp" = yes ; then
 	AC_CHECK_LIB(netsnmp,  netsnmp_ds_set_string,  [], [AC_MSG_ERROR([library 'netsnmp' is required for snmp support])])
 fi
 
+
```
Comment (hlinnaka) [inline configure.in]: spurious whitespace

File: configure.in [hunk 4]
```diff
@@ -224,7 +224,14 @@ PGAC_ARG_BOOL(enable, atomics, yes,
 # gpfdist
 #
 PGAC_ARG_BOOL(enable, gpfdist, yes,
-             [  --disable-gpfdist       do not use gpfdist])
+              [  --disable-gpfdist       do not use gpfdist],
+              [ if test "$with_openssl" = no || test "$with_openssl" = ""; then
```
Comment (hlinnaka) [inline configure.in]: Per discussion, adding the #ifdef USE_OPENSSL guards to gpfdist code would be nicer..

File: configure.in (hunk 2 above)
Comment (Chibin) [inline configure.in]: I'll remove this. gpfdist doesn't actually use this anymore and it uses
"#include <pg_config.h>"

File: configure.in (hunk 4 above)
Comment (Chibin) [inline configure.in]: Oh, I misunderstood and thought we would just want to use --with-openssl if gpfdist was enabled. My apologies... I'll take a look into this.

I also plan on not using "USE_SSL" and creating a new one "USE_OPENSSL" (hopefully?) for the gpfdist portion as USE_SSL causes problems for windows on our build side at the moment. Mainly due to the fact that the only potion we use OPENSSL for our windows is during gpfdist.

File: configure.in (hunk 3 above)
Comment (Chibin) [inline configure.in]: Removed, thanks

File: configure.in [hunk 5]
```diff
@@ -224,7 +224,8 @@ PGAC_ARG_BOOL(enable, atomics, yes,
 # gpfdist
 #
 PGAC_ARG_BOOL(enable, gpfdist, yes,
-             [  --disable-gpfdist       do not use gpfdist])
+              [  --disable-gpfdist       do not use gpfdist],
+              [ AC_DEFINE([USE_OPENSSL], 1, [ Define to build with (Open)SSL support (for gpfdist). ]) ])
 AC_SUBST(enable_gpfdist)
```
Comment (hlinnaka) [inline configure.in]: I was actually thinking of adding the required #ifdefs to gpfdist code, so that you could compile it without OpenSSL. Perhaps as a separate commit, but pushed together with this. Should be pretty straightforward.

File: configure.in [hunk 6]
```diff
@@ -904,9 +905,24 @@ AC_SUBST(with_libcurl)
...
+  APR_CONFIG=$with_apr_config
+fi
+# Check that gpdb_top is set correctly. When cross-compiling, AC_CHECK_FILE
+# cannot be used, so just trust that it is correct.
+if test "$cross_compiling" != "yes"; then
+  AC_CHECK_FILE($srcdir/src/backend/utils/misc/fstream/fstream.c, [], [
+    AC_MSG_ERROR([$srcdir variable must be set to the path of the GPDB sources])])
+fi
+
+case $host_os in
```
Comment (hlinnaka) [inline configure.in]: This should be unnecessary. There's a check for mingw near the top of configure.in, in the large "case $host_os" block

File: configure.in [hunk 7]
```diff
@@ -904,9 +905,24 @@ AC_SUBST(with_libcurl)
...
+[], [with_apr_config=yes])
+
+if test "$with_apr_config" = yes; then
+  AC_PATH_PROGS(APR_CONFIG, apr-1-config)
+else
+  APR_CONFIG=$with_apr_config
+fi
+# Check that gpdb_top is set correctly. When cross-compiling, AC_CHECK_FILE
+# cannot be used, so just trust that it is correct.
+if test "$cross_compiling" != "yes"; then
```
Comment (hlinnaka) [inline configure.in]: This gpdb_top check is not needed when gpfdist is directly in src/bin/gpfdist.

File: configure.in [hunk 8]
```diff
@@ -1104,17 +1120,38 @@ if test "$with_krb5" = yes ; then
...
+
+if test "$enable_gpfdist" = yes; then
+  # If the 'apr-1-config --link-ld' produced correct output, -lapr-1 is already
+  # in LIBS, hence AC_SEARCH_LIBS rather than AC_CHECK_LIB. (and the autoconf
+  # manual recommends always using AC_SEARCH_LIBS rather than AC_CHECK_LIB
+  # anyway)
+  AC_SEARCH_LIBS(apr_getopt_long, [apr-1], [], [AC_MSG_ERROR([libapr-1 is required for gpfdist])])
+  AC_SEARCH_LIBS(event_add, [event], [], [AC_MSG_ERROR([libevent is required for gpfdist])])
+
+  if test "$win32" != yes; then
```
Comment (hlinnaka) [inline configure.in]: I don't understand why this is conditional on win32. If you have libyaml installed in a Windows system, surely we would like to use it, and conversely, if you don't have libyaml installed on a Unix system, surely it should be possible to build without it.

File: configure.in [hunk 9]
```diff
@@ -1104,17 +1120,38 @@ if test "$with_krb5" = yes ; then
...
+    AC_SEARCH_LIBS(yaml_parser_initialize, [yaml], [], [AC_MSG_ERROR([libyaml is required for gpfdist])])
   fi
 fi
 
+if test "$with_openssl" = yes || test "$enable_gpfdist" = yes; then
+  AC_CHECK_LIB(crypto, CRYPTO_new_ex_data, [], [AC_MSG_ERROR([library 'crypto' is required for OpenSSL and/or gpfdist])])
+  AC_CHECK_LIB(ssl,    SSL_library_init, [], [AC_MSG_ERROR([library 'ssl' is required for OpenSSL and/or gpfdist])])
```
Comment (hlinnaka) [inline configure.in]: There are checks for these functions earlier in configure.in already

File: gpAux/Makefile [hunk 10]
```diff
@@ -67,13 +67,15 @@ sol10_sparc_32_CXX_LIB_DIR=/usr/sfw/lib
 BLD_CXX_LIB_DIR=$($(BLD_ARCH)_CXX_LIB_DIR)
 
 # help Windows builds find the cross-compiler and a pre-built curl
-include Makefile.windows
+ifeq ($(BLD_ARCH),win32)
```
Comment (hlinnaka) [inline gpAux/Makefile]: This seems like an unrelated change

File: configure.in (hunk 5 above)
Comment (Chibin) [inline configure.in]: > I was actually thinking of adding the required #ifdefs to gpfdist code, so that you could compile it without OpenSSL. Perhaps as a separate commit, but pushed together with this. Should be pretty straightforward.

I actually did that; I added ```#ifdefs USE_SSL``` within the gpfdist.c code, to compile with or without openssl.
I'll remove this USE_OPENSSL definition and I'll keep it as a separate commit as suggested.

File: configure.in (hunk 6 above)
Comment (Chibin) [inline configure.in]: Removing...

File: configure.in (hunk 8 above)
Comment (Chibin) [inline configure.in]: This was specifically for enabling transformations, but since win32 disables this functionality anyways. I can remove the conditional. I'll add a check for building with or without it.

File: gpAux/Makefile (hunk 10 above)
Comment (Chibin) [inline gpAux/Makefile]: We are just adding an explicit check for windows instead of the include file doing it for us.
This is related to gpfdist as the include file will add an .exe at the end of the filename.

**Comment by Chibin** [comment]
@hlinnaka Updated and split into 4 different commits based on the comments you gave. Thanks.

**Comment by hlinnaka** [comment]
Aside from the minor comments above, +1

**Comment by Chibin** [comment]
> Good catch. Did you get a compiler warning on this or something?

Found about this when I pulled the latest code and gpfdist crashed. The commit that caused the bug to show up was reverted [34826473dc2cf66592eb5bd207bbd6cce44ff90c] (src/backend/utils/misc/fstream/fstream.c:574),  but might as well fix the init issue here.

**Comment by Chibin** [comment]
Pushed, thanks for the review hlinnaka!
